### ISSUE TITLE
"struct is never used" when `Self` is used to name the type

### ISSUE THREAD

**Issue by jonas-schievink** [rust-lang/rust#30029]
``` rust
struct S;

impl Clone for S {
    fn clone(&self) -> Self { loop{} }
    // Doesn't warn:
    //fn clone(&self) -> S { loop{} }
}

fn main() {}
```

Prints a warning that `S` isn't used, but doesn't warn when naming `S` explicitly.

**Comment by Gankro** [comment]
It seems to me that *not* warning when it's used explicitly in a trait implementation is the real bug here.

**Comment by arielb1** [comment]
@Gankro

You can gain access to `S` via inference guessing.

**Comment by Mark-Simulacrum** [comment]
Closing in favor of (what I think is) a duplicate: https://github.com/rust-lang/rust/issues/18290.
